### PR TITLE
Ports Yogstation's Version of Disarm Rework (nerfs disarms)

### PR DESCRIPTION
**Pull request by Skrem7**
## About The Pull Request

Alters the disarm behavior already pulled from TG to match the balance changes made by Yogstation 13 coders when they merged TG's disarm rework four-ish years ago. See https://github.com/yogstation13/Yogstation/pull/9284

The following changes have been made:

1. Disarm interactions with tables and disposal bins have been removed
2. Shoving someone who is standing and wearing riot armor does nothing except move them back. No slow is applied, nor can a gun in their active hand be loosened and then dropped with two consecutive shoves. They still retain the inability to collide with mobs or have a mob colliding against them knock them down.
3. Shoving someone into a wall no longer knocks them down. Instead, their active hand item is dropped. Individuals wearing riot armor are unaffected.
4. Shoving someone who is prone stuns them for a reduced duration if they have melee armor located on their chest. The higher the armor, the more the three-second stun is reduced.
5. The knockdown for being shoved into someone has been reduced to a one-second knockdown on both parties, instead of three seconds for the shoved and one second for the person behind them.

Tested on a local server and the behavior seems to be working as intended.

## Why It's Good For The Game

1. Consistency with wall behavior and disarms being less "powerful". Considering some disposal bins go into space if I remember correctly, this is not good.
2. Makes riot armor appropriately potent at dealing with unarmed assailants.
3. Knockdowns are incredibly powerful, especially considering the follow-up disarm stun which guarantees an aggro grab. Unarmed individuals should not be able to so easily put someone at their mercy barring shoving into someone else or potentially slipping someone.
4. Makes it so melee armor helps against a melee-based attack. Makes sense.
5. Not sure why these were different values. There should certainly be a punishment for positioning in such a way that you line up for a knockdown, but the knockdown itself takes long enough to get up from that you can easily be kicked over and thoroughly savaged by an attacker.

While PvP is not the main focus of this server, stun combat is generally bad, nor should an unarmed, unprepared combatant be able to so easily overpower with better gear. The goal of the PR is to retain some of the behavior and strategy that could be utilized with disarm shoves while significantly reducing their power.

## Changelog

:cl:
balance: Disarms no longer knockdown individuals shoved into a wall. Instead, their active hand is disarmed.
balance: Riot armor now prevents ALL disarm intent disarms or slows; you can only be kicked over if already prone
balance: Armor reduces the duration of the paralyze if you're shoved while prone
/:cl:

### CODE REVIEW

**Comment by tmtmtl30** [comment]
the combo into an aggro grab is definitely more powerful than it should be, but a flat removal of the fundamental shovestun dynamic seems unnecessary? i've never much cared for the label "stun combat": i don't believe that somebody not being able to take an action for a mere 3 or 4 seconds during combat is necessarily a bad thing. equipment balance is also... complex, in that i think making equipment better generally makes player skill comparatively less important. that isn't bad on its own *either*, but ironically i think that fact is one of the things that causes people to *complain* about stun combat

**Comment by Skrem7** [comment]
> the combo into an aggro grab is definitely more powerful than it should be, but a flat removal of the fundamental shovestun dynamic seems unnecessary?

The fact of the matter is that anyone can overpower anyone else with their bare hands in a majority of pvp environments at the moment. That isn't skill, or preparation, or smart thinking. If you have nothing, the objectively best thing to do is to shove. That lacks critical depth. Punching someone to not only do actual, numerical damage and maybe knock them down VERSUS shoving them, slowing them, maybe getting their active hand to fumble which queues another instance of "who's better at scooping up gear" is much more interesting and open to player engagement and skill expression. Giving players without any items a hard stun is incredibly antithetical to the select few equipment items that CAN do that.

> i've never much cared for the label "stun combat": i don't believe that somebody not being able to take an action for a mere 3 or 4 seconds during combat is necessarily a bad thing.

I fundamentally disagree with how easy it is to get a leg up on someone with those three to four seconds, especially with how they're going to be set up to be knocked down again with whatever you just used, for the most part. With stun batons, this is its own issue, technically counteracted by battery, but that's its own discussion.

> equipment balance is also... complex, in that i think making equipment better generally makes player skill comparatively less important. that isn't bad on its own *either*, but ironically i think that fact is one of the things that causes people to *complain* about stun combat

People complain about stun combat because it innately prevents you from interacting with the game. People like to do things. I would rather be outgeared and outshot and still theoretically have had a chance to win with some shoddy 9mm than get rushed by a shove or a stun baton and just immediately ended because I don't have an answer that's so quickly and viscerally powerful, which stuns tend to be.

**Comment by tmtmtl30** [comment]
i'd be fine with a slight time nerf to shove stuns (looks like most servers use a 3-second stun instead of a 4-second one) and certain combos based on them (wall throwing and aggressive/neck grabs, in particular, but we also have some odd wind-up attacks that can be used as well). batons themselves are, also, an issue; they're a relic of station-focused combat design, where security *needs* to be able to trivially bring in resisting crewmembers for a variety of reasons. that's not that important here; a rebalance to batons, along with nerfs to the common unarmed attacks which chain off of shovestuns, seems to me like it'd fix the issue.

look, the problem i have with the objection that stun combat "prevents" you from playing the game is that dying *also does this*. in a one-on-one fight, getting stunned long enough for an enemy to cuff you is just as much of a loss as being beaten into crit. it feels different, because you're "not dead", but you've still lost the fight. i'm in favor of removing instant win buttons, but the shovestun *on its own* isn't one. without too much trouble we could ensure that there aren't combos that can turn it into one, while preserving the object-interaction dynamics it has that make unarmed combat interesting at all. and *if* it isn't an instant win, then a 3 or 4-second stun is not long enough to get up in arms over because it "prevents" you from playing.

i think you're somewhat over-exaggerating the difficulty in countering. shoving is not a foolproof method for winning a fight, because it inherently depends on positioning in a way that is highly unique to it. if you are being beaten by shoves, you can stay away from walls. positioning yourself defensively, before and during a fight, is a meaningful skill! players who lose to shoves still had chances to win: there was a fight beforehand where they allowed it to happen, or there *wasn't*, and the enemy took them by surprise when they hadn't placed themselves in a way that guarded against a pair of shoves.

yes, shoving is typically the best option in unarmed combat, but that doesn't mean it's an automatic or mindless decision. if you *aren't* able to get the opponent into a wall, you're wasting time you could be using to deal meaningful damage by pushing them around. it's only when you can get them into a wall and follow up with a shove-stun that it becomes effective, and that's not equally viable in all situations -- in an open room or against an opponent that is good at moving away from walls after shoves, it becomes totally ineffectual!

the importance of counterplay applies to active-hand-disarming attacks as well. the optimal counterplay for disarms at present is to only take out your weapons immediately before using them, sometimes via a hotkey, and then immediately reholster them, especially if they go on a cooldown. that strategy makes the chance to disarm somebody's active hand substantially less valuable, even assuming people don't get in the habit of just switching hands with x immediately before and after using any weapon.

**Comment by tmtmtl30** [comment]
i also asked any% about her thoughts on it, last night, and she asked me to post screenshots of her thoughts on the PR (her github password forgor):

image
image
image
image

**Comment by tmtmtl30** [comment]
blughghghghg. why did linux screenshot crop do that.................

**Comment by Skrem7** [comment]
> i'd be fine with a slight time nerf to shove stuns (looks like most servers use a 3-second stun instead of a 4-second one) and certain combos based on them (wall throwing and aggressive/neck grabs, in particular, but we also have some odd wind-up attacks that can be used as well). batons themselves are, also, an issue; they're a relic of station-focused combat design, where security _needs_ to be able to trivially bring in resisting crewmembers for a variety of reasons. that's not that important here; a rebalance to batons, along with nerfs to the common unarmed attacks which chain off of shovestuns, seems to me like it'd fix the issue.

This does have a time nerf on shove stuns in the form of the armor change. Aggressive grabs and batons could be tinkered too, but this PR fundamentally sought to target disarms first and foremost.

> look, the problem i have with the objection that stun combat "prevents" you from playing the game is that dying _also does this_. in a one-on-one fight, getting stunned long enough for an enemy to cuff you is just as much of a loss as being beaten into crit. it feels different, because you're "not dead", but you've still lost the fight. i'm in favor of removing instant win buttons, but the shovestun _on its own_ isn't one. without too much trouble we could ensure that there aren't combos that can turn it into one, while preserving the object-interaction dynamics it has that make unarmed combat interesting at all. and _if_ it isn't an instant win, then a 3 or 4-second stun is not long enough to get up in arms over because it "prevents" you from playing.

Dying is much more final. Playing against stuns means seeing your character whimper around on the ground while being perfectly capable of speech, while still technically being able to try things, to very little avail. I'll agree the shovestun isn't an instant win on its own, but you cannot deny its considerably greater strength than the other unarmed options.

> i think you're somewhat over-exaggerating the difficulty in countering. shoving is not a foolproof method for winning a fight, because it inherently depends on positioning in a way that is highly unique to it. if you are being beaten by shoves, you can stay away from walls. positioning yourself defensively, before and during a fight, is a meaningful skill! players who lose to shoves still had chances to win: there was a fight beforehand where they allowed it to happen, or there _wasn't_, and the enemy took them by surprise when they hadn't placed themselves in a way that guarded against a pair of shoves.

Staying away from walls innately requires you to not be indoors or be in open rooms. This is hilariously more practical criticism in station-based gameplay, because it's a lot easier to avoid maintenance for this type of thing. I also would like to point out the innate problem that still lies in the unarmed ability to effect such a powerful consequence. That is my issue here.

> yes, shoving is typically the best option in unarmed combat, but that doesn't mean it's an automatic or mindless decision. if you _aren't_ able to get the opponent into a wall, you're wasting time you could be using to deal meaningful damage by pushing them around. it's only when you can get them into a wall and follow up with a shove-stun that it becomes effective, and that's not equally viable in all situations -- in an open room or against an opponent that is good at moving away from walls after shoves, it becomes totally ineffectual!

It's not automatic or mindless, no. There's something to be said for conflict taking place on planets (where projectiles are going to always be king) versus on ships/stations (where it's a lot more sketchy). This seeks to actually maybe change up how you might fight in a station, rather than just shove, shove, shove.

> the importance of counterplay applies to active-hand-disarming attacks as well. the optimal counterplay for disarms at present is to only take out your weapons immediately before using them, sometimes via a hotkey, and then immediately reholster them, especially if they go on a cooldown. that strategy makes the chance to disarm somebody's active hand substantially less valuable, even assuming people don't get in the habit of just switching hands with x immediately before and after using any weapon.

You just expressed a potential expression of skill to counteract the unarmed assailant who is currently engaging you when you have a weapon. I don't see any problem with this.

**Comment by Skrem7** [comment]
> blughghghghg. why did linux screenshot crop do that.................

The fool.

As for the conversation, I'd argue that gear ON ships innately declares how much prowess a crew should have at something. There's something to be said for "planning" in getting gear for combat. Planning encompasses combat maneuvers as well as what you _actually_ bring into that combat. I prefer to emphasize the latter. If you force a fight when you aren't prepped for one, you shouldn't be surprised when the service officer pulls out her sidearm and blasts you.

Edit: Like, also consider the current cargo system too (I hate it I hate it I hate it). Want to jump someone? Buy a shotgun instead of an AKM.

**Comment by tmtmtl30** [comment]
ok here are the actual screenshots

image
image
image
image

**Comment by Zevotech** [comment]
A good close-quarters alternative to just shovestunning is throwing together a pipe shotgun and printing beanbags from the autolathe. Tested it, and it took me about 5 minutes on a beluga from spawning in to having a pipe shotgun. 2 shots for an unarmored target and 3 for a security officer to drop them for 4-6 seconds. Bonus points if you use rubbershot or saw off the shotgun to make it concealable. Ranged stun damage is accessible with five minutes of prep, even less for melee if you throw together a stunprod. 
Preparation is what I like to see. I don't want to tell myself "they're standing next to a wall and two tables, I've practically already won this fight" when I'm looking at someone pointing a pistol at me. I either don't have something ready and have to figure out how to get out of this situation, or I do have something ready and can think of how to use it. Thats fun. I wouldn't even be mad if I got kidnapped/disabled by someone conceal-carrying a stunprod, because they prepared for it. Getting my gun swiped and immediately unloaded into me for daring to stand within 2 tiles of John Wallshove isn't very fun or interesting to me.
I'd love to see this PR testmerged and tested full-on. I want to see how people adapt in a bar fight. People might actually use the stun given by smashing an empty bottle over someones head rather than just going for the guaranteed tableshove-tableslam stunlock that everyone knows and loves.

TLDR: Stun weapons are easy to make and people should use them more instead of relying on shoves. I want to see what deranged fighting game combos people come up with.

P.S: can you also remove the repeated-tableslam stunlock. If you want to keep someone stunned you should have a baton or disabler, not any random table.

**Comment by Skrem7** [comment]
Right. So.

Considering your need to wield most guns regarding recoil and shit now, holstering your gun until you fire is a really good way to disorient yourself.

You can still retreat from melee by shoving someone because they're still slowed. You can still disarm their melee weapon by shoving them against the wall, then run.

Calling knocking someone down into a hardstun "smarts and good positioning" is really a stretch. I suck at combat. I can still abuse this very easily with the few engagements I've gotten into. It's also allowed people who have no right to establish such a devastating level of control over my character with exactly nothing because of how quick, easy, and instant it is.

I would also like to point out that guns are still the _only_ item that can just be straight-up dropped with two shoves. If someone with a gun micros it out, that's... respect for what it is. It doesn't matter that they didn't drop the gun. They are still no longer holding the gun (not able to use it).

**Comment by Skrem7** [comment]
> P.S: can you also remove the repeated-tableslam stunlock. If you want to keep someone stunned you should have a baton or disabler, not any random table.

Would have to dig out what part of the code it's in and it wasn't intended to be part of this PR's scope, since this is already a rather drastic change.

**Comment by Zevotech** [comment]
True yeah, the fact that people have to wield the gun to properly aim it makes getting it out of someones hands a viable option for running away.
Or you could just steal the gun.

**Comment by Zevotech** [comment]
> Would have to dig out what part of the code it's in and it wasn't intended to be part of this PR's scope, since this is already a rather drastic change.

Yeah thats understandable. I'm just not a fan of it because it rewards being a one-trick in any combat situation with a table nearby.

**Comment by Skrem7** [comment]
> Yeah thats understandable. I'm just not a fan of it because it rewards being a one-trick in any combat situation with a table nearby.

I would be happy to take a look at it in a future PR (once I muster the motivation)! The same goes for batong, aggro-grab, and wall throw.

**Comment by Skrem7** [comment]
will cook something different i guess
